src/server/pkg/storage/fileset/priority_queue.go

```diff
@@ -0,0 +1,127 @@
+package fileset
+
+import (
+	"io"
+)
+
+type stream interface {
+	next() error
+	key() string
+}
+
+type priorityQueue struct {
+	queue []stream
+	size  int
+	ss    []stream
+}
+
+func newPriorityQueue(ss []stream) *priorityQueue {
+	return &priorityQueue{
+		queue: make([]stream, len(ss)+1),
+		ss:    ss,
+	}
+}
+
+func (pq *priorityQueue) iterate(f func([]stream, ...string) error) error {
+	for {
+		ss, err := pq.next()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+			return err
+		}
+		if err := f(ss, pq.peek()...); err != nil {
+			return err
+		}
+	}
+}
+
+func (pq *priorityQueue) key(i int) string {
+	return pq.queue[i].key()
+}
+
+func (pq *priorityQueue) empty() bool {
+	return pq.queue[1] == nil
+}
+
+func (pq *priorityQueue) insert(s stream) error {
+	// Get next in stream and insert it.
+	if err := s.next(); err != nil {
+		if err == io.EOF {
+			return nil
+		}
+		return err
+	}
+	pq.queue[pq.size+1] = s
+	pq.size++
+	// Propagate insert up the queue
+	i := pq.size
+	for i > 1 {
+		if pq.key(i/2) <= pq.key(i) {
+			break
+		}
+		pq.swap(i/2, i)
+		i /= 2
+	}
+	return nil
+}
+
+func (pq *priorityQueue) next() ([]stream, error) {
+	// Re-insert streams
+	if pq.ss != nil {
+		for _, s := range pq.ss {
+			if err := pq.insert(s); err != nil {
+				return nil, err
+			}
+		}
+	}
+	if pq.empty() {
+		return nil, io.EOF
+	}
+	ss := []stream{pq.queue[1]}
+	pq.fill()
+	// Keep popping streams off the queue if they have the same key.
+	for pq.queue[1] != nil && pq.key(1) == ss[0].key() {
+		ss = append(ss, pq.queue[1])
+		pq.fill()
+	}
+	pq.ss = ss
+	return ss, nil
+}
+
+func (pq *priorityQueue) peek() []string {
+	if pq.empty() {
+		return nil
+	}
+	return []string{pq.queue[1].key()}
+}
+
+func (pq *priorityQueue) fill() {
+	// Replace first stream with last
+	pq.queue[1] = pq.queue[pq.size]
+	pq.queue[pq.size] = nil
+	pq.size--
+	// Propagate last stream down the queue
+	i := 1
+	var next int
+	for {
+		left, right := i*2, i*2+1
+		if left > pq.size {
+			break
+		} else if right > pq.size || pq.key(left) <= pq.key(right) {
+			next = left
+		} else {
+			next = right
+		}
+		if pq.key(i) <= pq.key(next) {
+			break
+		}
+		pq.swap(i, next)
+		i = next
+	}
+}
+
+func (pq *priorityQueue) swap(i, j int) {
+	pq.queue[i], pq.queue[j] = pq.queue[j], pq.queue[i]
+}
```

Comment:
I think it'd be good to have some unittests for this.

Reply:
I agree, although I don't think it is high leverage right now. This is not on the critical path for getting the storage layer behind a feature flag, and can be done before 2.0 with the other testing improvements we have planned. This functionality is getting pretty heavily tested through the merge tests, but I can come up with some unit tests now if you want them before exposing the storage layer behind a feature flag.